iavl_tree_test.go

```diff
@@ -60,25 +60,32 @@ func TestVersionedRandomTree(t *testing.T) {
 func TestVersionedRandomTreeSmallKeys(t *testing.T) {
 	require := require.New(t)
 	tree := NewVersionedTree(100, db.NewMemDB())
+	singleVersionTree := NewVersionedTree(0, db.NewMemDB())
 	versions := 20
 	keysPerVersion := 50
 
 	for i := 1; i <= versions; i++ {
 		for j := 0; j < keysPerVersion; j++ {
 			// Keys of size one are likely to be overwritten.
-			tree.Set([]byte(cmn.RandStr(1)), []byte(cmn.RandStr(8)))
+			k := []byte(cmn.RandStr(1))
+			v := []byte(cmn.RandStr(8))
+			tree.Set(k, v)
+			singleVersionTree.Set(k, v)
 		}
 		tree.SaveVersion(uint64(i))
 	}
+	singleVersionTree.SaveVersion(1)
 
 	for i := 1; i < versions; i++ {
 		tree.DeleteVersion(uint64(i))
 	}
 
 	// After cleaning up all previous versions, we should have as many nodes
-	// in the db as in the current tree version.
+	// in the db as in the current tree version. The simple tree must be equal
```

Comment:
i like this test a lot.
can you make a variant that deletes the versions in a semi-random order rather than just in-order? to shake out any edge cases.

`for _, i := range rand.Perm(1,versions) { .. }`
(https://golang.org/pkg/math/rand/#Perm) can give you a random ordering

Reply:
Yes, good idea - also that is a cool function I didn't know about.